src/content/blog/2024-04-10-the-telemetry-tale.md

```diff
@@ -0,0 +1,105 @@
+
+### Architecture and Instrumentation
+
+> TODO: Image goes here
+
+The [number of transactions](https://github.com/interledger/rafiki/blob/d3be6b8d151d8cebc32b862e52a7bb678674d48e/packages/backend/src/open_payments/payment/outgoing/lifecycle.ts#L84-L90) is extracted from the Open Payments outgoing payment lifecycle and the value metric is handled by a [telemetry middleware](https://github.com/interledger/rafiki/blob/d3be6b8d151d8cebc32b862e52a7bb678674d48e/packages/backend/src/payment-method/ilp/connector/core/middleware/telemetry.ts) layer inside the ILP connector core.
+
+Before these metrics are sent to the Otel collectors we’ve implemented local privacy-preserving measures, including:
+- packetization which is inherent in the Interledger Protocol (ILP)
+- a currency conversion service that uses a custom AWS Lambda function that pulls daily exchange rates from configured sources and stores them in an Amazon S3 bucket
+- rounding data to a precision point determined by bucketing
+- local differential privacy in the form of adding Laplacian noise to anonymize transaction data
+
+Finally, we grappled with the dilemma of disentangling playground data from our test network and live production data. The [Test Network](https://github.com/interledger/testnet) (Testnet) is an open Interledger network using test money within example wallet and e-commerce applications. Rafiki also provides a [local playground](https://github.com/interledger/rafiki/tree/main/localenv) to experiment with. We’d like to monitor activity within the testing environments to gauge interest in the Interledger Protocol (ILP) and identify usage patterns by new participants. We’ve opted to provide users the choice between the test network (testnet) and the live network (livenet). To support this dual-environment approach, the infrastructure for telemetry was effectively expanded to include two separate services: one dedicated to collecting and managing telemetry data from the test network and another for the live network. This setup ensures that data from each environment is handled independently.
+
+We’re using an AWS ECS Fargate deployment for our Telemetry Cluster. The cluster holds services for our two environments: livenet and testnet, both of them having a Network Load Balancer (NLB) in order to load-balance over their own set of ECS task replicas. By using an NLB instead of the Application Load Balancer (ALB), we are not taking the hit of parsing the L7 ([OSI Model](https://en.wikipedia.org/wiki/OSI_model)) protocols and we do not have to do TLS termination at load balancer level. In our case, We just need to pass the data received to Prometheus, and application-layer routing decisions are not needed. Acting on L4 is of greater convenience, as all we need is the TCP packets being redirected to one of the running ECS tasks.
```

Comment:
```suggestion
We’re using an AWS ECS Fargate deployment for our Telemetry Cluster. The cluster holds services for our two environments: livenet and testnet, both of them having a Network Load Balancer (NLB) in order to load-balance over their own set of ECS task replicas. By using an NLB instead of the Application Load Balancer (ALB), we are not taking the hit of parsing the L7 ([OSI Model](https://en.wikipedia.org/wiki/OSI_model)) protocols and we do not have to do TLS termination at load balancer level. In our case, we just need to pass the data received to Prometheus, and application-layer routing decisions are not needed. Acting on L4 is of greater convenience, as all we need is the TCP packets being redirected to one of the running ECS tasks.
```

Reply:
Also, (personally) the part:

> By using an NLB instead of the Application Load Balancer (ALB), we are not taking the hit of parsing the L7 ([OSI Model](https://en.wikipedia.org/wiki/OSI_model)) protocols and we do not have to do TLS termination at load balancer level. In our case, we just need to pass the data received to Prometheus, and application-layer routing decisions are not needed. Acting on L4 is of greater convenience, as all we need is the TCP packets being redirected to one of the running ECS tasks.

could use from a bit more context? A lot of acronyms all at once in here 😅

Reply:
There is a misunderstanding.
It's not the case of a load balancer that knows where to route the data based on metric introspection/metadata, and then figures if it should send it to livenet or testnet.

But 2 load balancers, already assigned to one of the two environments. So the testnet load balancer just distributes metrics evenly across multiple testnet otel collectors, and the livenet load balancer just distributes the metrics evenly across multiple livenet otel collectors.

By having a LIVENET flag in the rafiki deployment, the rafiki instance already knows if it is sending to livenet or testnet. That distinction is done at client side, not in the load balancers.

"Our telemetry cluster is deployed on AWS, utilizing the ECS Fargate service to cater to two distinct operational environments: the live network environment (livenet) and the test network environment (testnet). To ensure efficient distribution of incoming network traffic across ECS task replicas within each environment, we employ Network Load Balancers (NLBs)."

The terms here might be a bit misleading, I will come back to this and try to rephrase

Reply:
Thanks so much for that clarification @beniaminmunteanu. How about this:

Our telemetry cluster is deployed on AWS, utilizing the ECS Fargate service to support two distinct operational environments: the live network environment (livenet) and the test network environment (testnet). To ensure efficient distribution of incoming network traffic across ECS task replicas within each environment, we employ Network Load Balancers (NLBs). The presence of a dedicated load balancer for each environment enables us to opt for NLBs over Application Load Balancers, effectively sidestepping the complexities associated with parsing application layer protocols (Layer 7 of the OSI model).

This strategic choice significantly impacts how we handle Transport Layer Security (TLS). In the case where the load balancer determines which environment the data should be sent to, managing TLS termination at the load balancer level would necessitate the decryption of incoming HTTPS traffic before passing it on to the backend servers in an unencrypted HTTP form. Such a process demands CPU resources for both decryption and HTTP traffic parsing, leading to increased resource usage and higher operational costs.

However, since the differentiation between environments (livenet and testnet) is managed by the Rafiki client and not within the load balancers, we leverage NLBs which operate at the transport layer. This approach removes the need for TLS termination at the load balancer, allowing us to establish a TLS secure channel directly with the services behind the load balancer (the respective Prometheus instance), rather than with the load balancer itself. Rafiki sends test data directly to the testnet environment and live data to the livenet environment. This configuration optimizes our resource utilization, leading to a more efficient and cost-effective telemetry infrastructure.

Reply:
"This strategic choice significantly impacts how we handle Transport Layer Security (TLS). In the case where the load balancer determines which environment the data should be sent to based on the data itself, it would have to act as a TLS terminator, decrypting the incoming GRPC/HTTPS traffic and inspecting it before passing it on to the selected backend server. Such a process demands CPU resources for both decryption and parsing of the traffic before passing it on to the backend servers, leading to increased resource usage and higher operational costs"

I made some small adjustments here: 

- Wanted to make it clearer that the case for L7 would be when the distinction of environments is done by using information from the data itself. 

- Rephrased the Managing TLS termination part a bit

- Also, the communication between the load balancer and a backend server does not necessarily need to be unencrypted, more than often it ends up being encrypted as well, so I removed the "unencrypted HTTP" part

- Mentioned GRPC, since in our case that is the actual traffic, not HTTPS


Also, in the next paragraph:
"This approach removes the need for TLS termination at the load balancer, allowing us to establish a TLS secure channel directly with the services behind the load balancer (the respective Prometheus instance)," <--this part

The thing we connect to behind the load balancer is an ADOT collector, not the prometheus instance.
We are load balancing between multiple ADOT collectors. (more specifficaly, ECS Tasks that run an ADOT collector container).
So maybe just remove the "(the respective Prometheus instance)" and it's fine.